src/librustdoc/html/static/js/search.js

```diff
@@ -131,35 +365,43 @@ window.initSearch = function(rawSearchIndex) {
      * @param  {[type]} filterCrates [Crate to search in if defined]
      * @return {[type]}              [A search index of results]
      */
-    function execQuery(query, searchWords, filterCrates) {
-        function itemTypeFromName(typename) {
-            for (var i = 0, len = itemTypes.length; i < len; ++i) {
-                if (itemTypes[i] === typename) {
-                    return i;
-                }
+    function execQuery(queryInfo, searchWords, filterCrates) {
+        // The goal of this new way of storing results is to prevent having too much elements to
+        // sort.
```

Comment:
I tried to write good comments but if you have ideas for improvements, I'm very glad to apply them! :)

> Also, this seems separate from the parser improvements. Could it be done as a separate PR?

This is something that is under debate I think here: the parser alone isn't of much use with the current search because it can basically checks the three cases:
 1. Is it an exact match? (using double quotes)
 2. Is it a weird query with the return arrow? (with `->`)
 3. If none of the 2 previous, we match the whole string.

I can just make the parser update though, just that it would be greatly overkill but if that makes things simpler for review, totally worth it!

Reply:
There's not much point in adding new functionality to the parser without adding the new functionality to the search engine. I agree with that.

But this changed internal representation of the results (where it's linked by `Math.floor(levenshtein)` instead of a flat list) seems like a pure optimization that's not related to the functionality improvements. If the added functionality means it's too slow without the optimization, then the optimization could be extracted from this PR and landed first.

Reply:
To make things simpler, I'll first just add the parser from this PR into a separate PR. Like that it'll make the PR smaller. @jyn514 rightfully noted that it extended the search capabilities and that it should discussed beforehand, so I'll see with the rustdoc team if we want this extension and what are the steps to do so.

As for the optimization: it was useful at the beginning when I was debugging but it should be put in a PR on its own. Things would have been much simpler if JS had a `BTreeMap` type, but unfortunately it doesn't...